bin/src/_main.rs

```diff
@@ -367,28 +383,42 @@ pub async fn _main(
             }
             _ => false,
         },
-        |(watched_dirs, rules, lookback, offset_state, fo_state_handles)| {
-            let mut initial_offsets: Option<HashMap<FileId, SpanVec>> = None;
-
-            if let Some(offset_state) = offset_state {
-                match offset_state.offsets() {
-                    Ok(offsets) => {
-                        initial_offsets =
-                            Some(offsets.into_iter().map(|fo| (fo.key, fo.offsets)).collect());
-                    }
-                    Err(e) => warn!("couldn't retrieve offsets from agent state, {:?}", e),
-                }
-            }
+        |(watched_dirs, rules, lookback, fo_state_handles, fs_offsets)| {
+            let watched_dirs = watched_dirs.clone();
+            let rules = rules.clone();
+            let lookback = lookback.clone();
+            let deletion_ack_sender = deletion_ack_sender.clone();
+            let fo_state_handles = fo_state_handles.clone();
+            let fs_offsets = fs_offsets.clone();
+            async move {
+                let tailer = tail::Tailer::new(
+                    watched_dirs,
+                    rules,
+                    lookback,
+                    Some(fs_offsets.lock().await.clone()),
+                    fo_state_handles,
+                    deletion_ack_sender,
+                );
 
-            let tailer = tail::Tailer::new(
-                watched_dirs.clone(),
-                rules.clone(),
-                lookback.clone(),
-                initial_offsets,
-                fo_state_handles.clone(),
-                deletion_ack_sender.clone(),
-            );
-            async move { tail::process(tailer).expect("Failed to create FS Tailer") }
+                tail::process(tailer)
+                    .expect("Failed to create FS Tailer")
+                    .filter(move |line| {
+                        let mut pair = (None, None);
+                        if let Ok(line) = line {
+                            pair = (line.get_key(), line.get_offset());
+                        }
+
+                        let fs_offsets = fs_offsets.clone();
+                        async move {
+                            if let (Some(key), Some(offsets)) = pair {
+                                let mut span_vec = SpanVec::new();
+                                span_vec.insert(Span::try_from(offsets).unwrap());
```

Comment:
this is the only place where we read offsets from DB?

Reply:
Offsets from the DB are only used once on agent startup as seen in L364. All offsets thereafter are updated as lines come through.

Reply:
fs_offsets.clone() gets updated?

Reply:
`fs_offsets` gets updated from each line that passes through. Each time the tailer is restarted, the new tailer gets a clone of fs_offsets so it starts with the the most up-to-date offsets.

Reply:
i mean - why do you need clone() in the first place?